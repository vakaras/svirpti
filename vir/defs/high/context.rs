pub trait Context {}

pub type Sort = super::typ::Type;
pub type Expression = super::expression::Expression;
pub type Statement = super::statement::Statement;
crate::derive_string_symbol!(UninterpretedSortSymbol);
crate::derive_string_symbol!(FieldNameSymbol);
crate::derive_string_symbol!(AdtNameSymbol);
crate::derive_string_symbol!(VariableSymbol);
crate::derive_string_symbol!(FunctionSymbol);
crate::derive_string_symbol!(AxiomNameSymbol);
crate::derive_string_symbol!(LabelSymbol);
