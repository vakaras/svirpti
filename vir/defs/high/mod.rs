pub mod context;
pub mod declaration;
pub mod expression;
pub mod parse;
pub mod program;
pub mod statement;
pub mod typ;

pub use context::*;
pub use declaration::VariableDeclaration;
pub use expression::{
    BinaryOperation, BinaryOperationKind, BoundedVariableDecl, Conditional, Constant,
    FunctionApplication, Quantifier, QuantifierKind, Trigger, UnaryOperation, UnaryOperationKind,
    Variable,
};
pub use program::{BasicBlock, BasicBlockId, ProcedureDeclaration, ProgramFragment};
pub use statement::{Assert, Assume, Havoc, Assign};
pub use typ::{DomainType, Type};
