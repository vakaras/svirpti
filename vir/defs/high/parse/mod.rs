pub mod declaration;
pub mod expression;
pub mod program;
pub mod sort;
pub mod statement;

pub use program::ProgramFragment;
