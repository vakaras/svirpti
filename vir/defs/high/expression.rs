use super::context::*;

vir_include! { expression =>
    use Variable;
    use Constant;
    use UnaryOperation;
    use UnaryOperationKind;
    use BinaryOperation;
    use BinaryOperationKind;
    use Conditional;
    use Quantifier;
    use QuantifierKind;
    use Trigger;
    use BoundedVariableDecl;
    use FunctionApplication;
    derive PartialEq, Eq, Debug, Clone, serde::Serialize, serde::Deserialize;
}
vir_include! { expression::helpers =>
    use VariableHelpers;
    use ConstantHelpers;
    use UnaryOperationHelpers;
    use BinaryOperationHelpers;
    use QuantifierHelpers;
    use FunctionApplicationHelpers;
}
pub use crate::common::expression::{
    BinaryOperationHelpers, ConstantHelpers, FunctionApplicationHelpers, QuantifierHelpers,
    UnaryOperationHelpers, VariableHelpers,
};
vir_include! { expression::evaluation =>
    use Variable;
    use Constant;
    use UnaryOperation;
    use BinaryOperation;
    use Conditional;
    use Quantifier;
    use FunctionApplication;
}
pub use crate::common::expression::SyntacticEvaluation;
vir_include! { expression::display =>
    use Variable;
    use Constant;
    use UnaryOperation;
    use BinaryOperation;
    use BinaryOperationKind;
    use Conditional;
    use Quantifier;
    use QuantifierKind;
    use BoundedVariableDecl;
    use Trigger;
    use FunctionApplication;
}

#[derive(PartialEq, Eq, Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum Expression {
    Variable(Variable),
    Constant(Constant),
    UnaryOperation(UnaryOperation),
    BinaryOperation(BinaryOperation),
    Conditional(Conditional),
    Quantifier(Quantifier),
    FunctionApplication(FunctionApplication),
}

impl SyntacticEvaluation for Expression {
    fn is_true(&self) -> bool {
        match self {
            Expression::Variable(expr) => expr.is_true(),
            Expression::Constant(expr) => expr.is_true(),
            Expression::UnaryOperation(expr) => expr.is_true(),
            Expression::BinaryOperation(expr) => expr.is_true(),
            Expression::Conditional(expr) => expr.is_true(),
            Expression::Quantifier(expr) => expr.is_true(),
            Expression::FunctionApplication(expr) => expr.is_true(),
        }
    }
    fn is_false(&self) -> bool {
        match self {
            Expression::Variable(expr) => expr.is_false(),
            Expression::Constant(expr) => expr.is_false(),
            Expression::UnaryOperation(expr) => expr.is_false(),
            Expression::BinaryOperation(expr) => expr.is_false(),
            Expression::Conditional(expr) => expr.is_false(),
            Expression::Quantifier(expr) => expr.is_false(),
            Expression::FunctionApplication(expr) => expr.is_false(),
        }
    }
}

impl std::fmt::Display for Expression {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Expression::Variable(expr) => expr.fmt(f),
            Expression::Constant(expr) => expr.fmt(f),
            Expression::UnaryOperation(expr) => expr.fmt(f),
            Expression::BinaryOperation(expr) => expr.fmt(f),
            Expression::Conditional(expr) => expr.fmt(f),
            Expression::Quantifier(expr) => expr.fmt(f),
            Expression::FunctionApplication(expr) => expr.fmt(f),
        }
    }
}
