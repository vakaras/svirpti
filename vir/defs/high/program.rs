use super::context::*;
use super::declaration::*;
use index_vec::IndexVec;

vir_include! { cfg =>
    use GuardedBasicBlock as BasicBlock;
    use Ids;
    use BasicBlockWithSuccessors;
    derive PartialEq, Eq, Debug, Clone, serde::Serialize, serde::Deserialize;
}

#[derive(PartialEq, Eq, Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ProgramFragment {
    pub sorts: Vec<UninterpretedSortDeclaration>,
    pub axioms: Vec<AxiomDeclaration>,
    pub functions: Vec<FunctionDeclaration>,
    pub procedure: ProcedureDeclaration,
}

#[derive(PartialEq, Eq, Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ProcedureDeclaration {
    pub variables: IndexVec<VariableId, VariableDeclaration>,
    pub basic_blocks: IndexVec<BasicBlockId, BasicBlock>,
}

index_vec::define_index_type! {
    pub struct VariableId = usize;
}

impl crate::common::cfg::Cfg for ProcedureDeclaration {
    type BasicBlock = BasicBlock;
    fn basic_blocks(&self) -> &IndexVec<BasicBlockId, BasicBlock> {
        &self.basic_blocks
    }
}

impl std::fmt::Display for ProgramFragment {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "program {{")?;
        writeln!(f, "  sorts: {:?}", self.sorts)?;
        writeln!(f, "  functions: {:?}", self.functions)?;
        writeln!(f, "  axioms: {:?}", self.axioms)?;
        writeln!(f, "  procedure:")?;
        writeln!(f, "    variables:")?;
        for variable in &self.procedure.variables {
            writeln!(f, "      {}: {}", variable.name, variable.sort)?;
        }
        writeln!(f, "    blocks:",)?;
        for (id, basic_block) in self.procedure.basic_blocks.iter_enumerated() {
            writeln!(f, "      [{:?}] {}:", id, basic_block.label)?;
            writeln!(f, "        guard: {}", basic_block.guard)?;
            for statement in &basic_block.statements {
                writeln!(f, "        {}", statement)?;
            }
            writeln!(f, "        successors: {:?}", basic_block.successors)?;
        }
        writeln!(f, "}}")
    }
}