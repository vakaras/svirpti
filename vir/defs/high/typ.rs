use super::context::*;

vir_include! { typ =>
    use Type;
    use DomainType;
    derive PartialEq, Eq, Debug, Clone, serde::Serialize, serde::Deserialize;
}
vir_include! { typ::display =>
    use Type;
    use DomainType;
}
