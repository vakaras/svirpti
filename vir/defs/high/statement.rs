use super::context::*;
use super::expression::Variable;

vir_include! { statement =>
    use Assert;
    use Assume;
    use Havoc;
    use Assign;
    derive PartialEq, Eq, Debug, Clone, serde::Serialize, serde::Deserialize;
}
vir_include! { statement::helpers =>
    use AssumeAssertHelpers;
}
pub use crate::common::statement::AssumeAssertHelpers;
vir_include! { statement::display =>
    use Assert;
    use Assume;
    use Havoc;
    use Assign;
}

#[derive(PartialEq, Eq, Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum Statement {
    Assume(Assume),
    Assert(Assert),
    Havoc(Havoc),
    Assign(Assign),
}

impl std::fmt::Display for Statement {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Statement::Assert(statement) => statement.fmt(f),
            Statement::Assume(statement) => statement.fmt(f),
            Statement::Havoc(statement) => statement.fmt(f),
            Statement::Assign(statement) => statement.fmt(f),
        }
    }
}