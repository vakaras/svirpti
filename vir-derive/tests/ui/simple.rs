use svirpti_vir_derive::vir_high;
use svirpti_vir::high::ProgramFragment;

fn empty() {
    let _ = vir_high!{

    };
}

fn no_locals() {
    let _ = vir_high!{
        procedure { }
    };
}

fn minimal() {
    let _: ProgramFragment = vir_high!{
        procedure {
            locals {}
        }
    };
}

fn exit_basic_block() {
    let _: ProgramFragment = vir_high!{
        procedure {
            locals {}
            exit { }
        }
    };
}

fn single_basic_block() {
    let _: ProgramFragment = vir_high!{
        procedure {
            locals {}
            bb1 {
                guard true;
                goto { exit }
            }
        }
    };
}

fn missing_basic_block() {
    let _: ProgramFragment = vir_high!{
        procedure {
            locals {}
            bb1 {
                guard true;
                goto { bb2 }
            }
        }
    };
}

fn diamond_basic_block() {
    let _: ProgramFragment = vir_high!{
        procedure {
            locals {}
            bb1 {
                guard true;
                goto { bb2, bb3 }
            }
            bb2 {
                guard true;
                goto { bb4 }
            }
            bb3 {
                guard true;
                goto { bb4 }
            }
            bb4 {
                guard true;
                goto { exit }
            }
        }
    };
}

fn main() {}