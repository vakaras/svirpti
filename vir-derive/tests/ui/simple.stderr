error: missing procedure
 --> $DIR/simple.rs:5:13
  |
5 |       let _ = vir_high!{
  |  _____________^
6 | |
7 | |     };
  | |_____^
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected end of input, expected `locals`
  --> $DIR/simple.rs:12:21
   |
12 |         procedure { }
   |                     ^

error: `exit` is reserved for exiting procedure
  --> $DIR/simple.rs:28:13
   |
28 |             exit { }
   |             ^^^^

error: no block with this label
  --> $DIR/simple.rs:51:24
   |
51 |                 goto { bb2 }
   |                        ^^^
